pkg/controller/master/virtualmachine/vm_controller.go

```diff
@@ -240,61 +164,38 @@ func (h *VMController) StoreRunStrategy(_ string, vm *kubevirtv1.VirtualMachine)
 	return vm, nil
 }
 
-// OnVMRemove handle related PVC and delete related VMBackup snapshot
-func (h *VMController) OnVMRemove(vm *kubevirtv1.VirtualMachine) (*kubevirtv1.VirtualMachine, error) {
+// cleanupPVCAndSnapshot handle related PVC and delete related VMBackup snapshot
+func (h *VMController) cleanupPVCAndSnapshot(_ string, vm *kubevirtv1.VirtualMachine) (*kubevirtv1.VirtualMachine, error) {
+	if vm == nil {
+		return vm, nil
+	}
 
-	if err := h.unsetOwnerOfPVCs(vm); err != nil {
+	if err := h.removeVMBackupSnapshot(vm); err != nil {
 		return vm, err
 	}
 
-	if err := h.removeVMBackupSnapshot(vm); err != nil {
+	if err := h.removePVCInRemovedPVCsAnnotation(vm); err != nil {
 		return vm, err
 	}
 
-	// remove harvester finalizers
-	vmObj := vm.DeepCopy()
-	util.RemoveFinalizer(vmObj, harvesterUnsetOwnerOfPVCsFinalizer)
-	util.RemoveFinalizer(vmObj, oldWranglerFinalizer) // post upgrade, these are not being cleaned by wrangler so need to be managed by us
-	return h.vmClient.Update(vmObj)
+	return vm, nil
 }
 
-// unsetOwnerOfPVCs erases the target VirtualMachine from the owner of the PVCs in annotation.
-func (h *VMController) unsetOwnerOfPVCs(vm *kubevirtv1.VirtualMachine) error {
-	var (
-		pvcNamespace = vm.Namespace
-		pvcNames     = getPVCNames(&vm.Spec.Template.Spec)
-		removedPVCs  = getRemovedPVCs(vm)
-	)
-	for _, pvcName := range pvcNames.List() {
-		var pvc, err = h.pvcCache.Get(pvcNamespace, pvcName)
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				// NB(thxCode): ignores error, since this can't be fixed by an immediate requeue,
-				// and also doesn't block the whole logic if the PVC has already been deleted.
-				continue
-			}
-			return fmt.Errorf("failed to get PVC(%s/%s): %w", pvcNamespace, pvcName, err)
-		}
-
-		if err := unsetBoundedPVCReference(h.pvcClient, pvc, vm); err != nil {
-			return fmt.Errorf("failed to revoke VitrualMachine(%s/%s) as PVC(%s/%s)'s owner: %w",
-				vm.Namespace, vm.Name, pvcNamespace, pvcName, err)
+// removePVCInRemovedPVCsAnnotation cleanup PVCs which users want to remove with the VM.
+func (h *VMController) removePVCInRemovedPVCsAnnotation(vm *kubevirtv1.VirtualMachine) error {
+	for _, pvcName := range getRemovedPVCs(vm) {
+		pvcName = strings.TrimSpace(pvcName)
```

Comment:
Moved `TrimSpace` to `getRemovedPVCs` function.

We have to avoid error data from users. I will use https://github.com/harvester/harvester/issues/5456 to validate annotation value in webhook.